### ISSUE TITLE
Start before stopping (downtime less rolling deployment) not working on some services

### ISSUE THREAD

**Issue by Tigraine** [rancher/rancher#6593]
Hi,
I we running rancher v1.1.4 on Docker 1.10.3-0~trusty on Ubuntu 14.04 LTS in AWS EC2.
It's a HA setup with 3 Servers with Cattle.

On most services (after setting up health checks) we can upgrade a service with "Start before stopping" where Rancher will first wait for the new container to become healthy before it shuts down the old container. This way we get no downtimes when deploying a new version of the service - this is great and the expected behavior. (Btw: Why is the start before stop check not on by default - and how can we set this default?)

But on the same Rancher setup some services (with identical health checks etc) will just shut down the old container even though the new one is still in initializing state. Is this a known issue? How can this work on one service and not on another in the same cluster.

Thanks for your help. 
Greetings Daniel

**Comment by Tigraine** [comment]
Update on the situation: We just upped the scale on this service to 2 and tried upgrade again.

This time it waited for the third container to turn healthy, then started the 4th and immediately shut down the first 2. Better, but still not ideal as preferably i'd like 4 containers running and healthy before it is killing my old ones.

**Comment by aemneina** [comment]
hey @Tigraine how many hosts do you have, and are you binding the service to a port? Also do you have an example docker/rancher-compose.yml files, for your stack/service definitions, we can take a look at?

**Comment by Tigraine** [comment]
Hi @aemneina thanks for looking into this. Here is the rancher compose file in question:

```
analytics-resource-service-canary-internal:
  scale: 2
  health_check:
    port: 8080
    interval: 2000
    initializing_timeout: 180000
    unhealthy_threshold: 3
    strategy: none
    response_timeout: 2000
    request_line: GET "/health" "HTTP/1.0"
    healthy_threshold: 2
``` 

We are running 8 hosts at the moment (2 of these are gateways we only schedule Load Balancers to and 6 hosts with the services). The service is not bound to a port. There is a Load Balancer in front of the service on dedicated hosts.

If the service is at scale 1 it will shut down the old container immediately once the new one is started - won't wait for the health check to actually pass. In the meantime the service won't respond for the rest of the time until the health check succeeds again (our service takes around 30 secs to start up)
Downtime in this case ~20-30 seconds.

If I scale it to 2 first it will will do the following:
- Start a third container
- Once the third container is initializing it will shut down one of the 2 old containers (at this point 2-3 requests timed out but nothing too serious
- It will wait for the health checks on the third container to pass then it will start a 4th container and once that's initializing it will immediately stop the last remaining old container.

With scale 2 it is also showing the same bad behavior it's just mitigated by having one old host remaining that can serve requests. (And weirdly waiting for the health checks to pass before continuing the upgrade - but not when stopping old services?)

In both cases I checked "Start before stopping".

**Comment by aemneina** [comment]
@Tigraine looks like this is a known issue fixed in 1.2.0-pre4-rc3 (i think the latest rc's are not that stable). See here: https://github.com/rancher/rancher/issues/5826 I think you might want to wait for the v1.2.0 GA before rolling this out to production, but doesnt hurt to validate the use case for yourself.

**Comment by aemneina** [comment]
Will close this out, if this isnt fixed in 1.2.0GA feel free to reopen or open a new issue.

**Comment by Tigraine** [comment]
Thanks for the info @aemneina .
